app/Models/ContestEntry.php

```diff
@@ -52,6 +53,53 @@ public function votes()
         return $this->hasMany(ContestVote::class);
     }
 
+    public function scopeForBestOf(Builder $query, User $user, string $ruleset, ?string $variant = null): Builder
+    {
+        $beatmapsetIdsQuery = Beatmap::select('beatmapset_id')
+            ->where('playmode', Beatmap::MODES[$ruleset])
+            ->whereIn('beatmap_id', $user->beatmapPlaycounts()->select('beatmap_id'));
+
+        if ($ruleset === 'mania' && $variant !== null) {
+            if ($variant === 'nk') {
+                $beatmapsetIdsQuery->whereNotIn('diff_size', [4, 7]);
+            } else {
+                $keys = match ($variant) {
+                    '4k' => 4,
+                    '7k' => 7,
+                };
+                $beatmapsetIdsQuery->where('diff_size', $keys);
+            }
+        }
+
+        return $query->whereIn('entry_url', $beatmapsetIdsQuery);
+    }
+
+    // TODO: switch Contest param with casted extra_options object.
+    public function scopeWithScore(Builder $query, Contest $contest): Builder
```

Comment:
looking again I don't know if it's better here compared to inside `Contest` as it depends on passing `$contest` which could very well be unrelated. It also looks weird when accessed from `$contest` - `$contest->entries()->withScore($contest)`.

I think I'd rather have `entriesWithScore()` in the `Contest` itself...

Reply:
Not all entries in the contest need the score when viewing a single entry, and I'd rather split all the types of scopes into separate scopes/functions instead of overloading everything into the same fields, but that's not happening until  I get around to replacing the existing contests coffeescript.